File: Household Budget Calculator/[C#]-Household Budget Calculator/C#/Household budget calculator CS/properties.cs
//using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.ComponentModel;
using System.Globalization;
using System.Drawing.Design;
using Household_budget_calculator_CS.expanding_property_classes;
using Household_budget_calculator_CS.expanding_property_classes.TypeEditors;


namespace Household_budget_calculator_CS
{
    [System.Serializable()]
    public class properties
    {
        public event propertiesChangedEventHandler propertiesChanged;
        public delegate void propertiesChangedEventHandler(propertiesChangedEventArgs e);

        public properties()
        {
            removeEventHandlers();
            addEventHandlers();
        }

        public void addEventHandlers()
        {
            _children.propertiesChanged += expandable_propertiesChanged;
            _finance.propertiesChanged += expandable_propertiesChanged;
            _houseHold.propertiesChanged += expandable_propertiesChanged;
            _insurance.propertiesChanged += expandable_propertiesChanged;
            _leisure.propertiesChanged += expandable_propertiesChanged;
            _otherBills.propertiesChanged += expandable_propertiesChanged;
            _regularBills.propertiesChanged += expandable_propertiesChanged;
            _transport.propertiesChanged += expandable_propertiesChanged;
        }

        public void removeEventHandlers()
        {
            _children.propertiesChanged -= expandable_propertiesChanged;
            _finance.propertiesChanged -= expandable_propertiesChanged;
            _houseHold.propertiesChanged -= expandable_propertiesChanged;
            _insurance.propertiesChanged -= expandable_propertiesChanged;
            _leisure.propertiesChanged -= expandable_propertiesChanged;
            _otherBills.propertiesChanged -= expandable_propertiesChanged;
            _regularBills.propertiesChanged -= expandable_propertiesChanged;
            _transport.propertiesChanged -= expandable_propertiesChanged;
        }

        private decimal _net;
        [CategoryAttribute("(Monthly Income)"), RefreshProperties(RefreshProperties.All), DisplayName("Net monthly pay")]
        public decimal net
        {
            get { return _net; }
            set
            {
                _net = value;
                totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2");
            }
        }

        private decimal _pension;
        [CategoryAttribute("(Monthly Income)"), RefreshProperties(RefreshProperties.All), DisplayName("Pension")]
        public decimal pension
        {
            get { return _pension; }
            set
            {
                _pension = value;
                totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2");
            }
        }

        private decimal _otherIncome1;
        [CategoryAttribute("(Monthly Income)"), RefreshProperties(RefreshProperties.All), DisplayName("Income from Savings + Investments")]
        public decimal otherIncome1
        {
            get { return _otherIncome1; }
            set
            {
                _otherIncome1 = value;
                totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2");
            }
        }

        private decimal _otherIncome2;
        [CategoryAttribute("(Monthly Income)"), RefreshProperties(RefreshProperties.All), DisplayName("Other Income")]
        public decimal otherIncome2
        {
            get { return _otherIncome2; }
            set
            {
                _otherIncome2 = value;
                totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2");
            }
        }

        private string _totalIncome;
        [CategoryAttribute("(Monthly Income)"), RefreshProperties(RefreshProperties.All), ReadOnlyAttribute(true), DisplayName("Total Income"), Editor(typeof(TypeEditor0), typeof(UITypeEditor))]
        public string totalIncome
        {
            get { return _totalIncome; }
            set
            {
                _totalIncome = value;
                decimal decValue = default(decimal);
                decimal.TryParse(value, NumberStyles.Currency, CultureInfo.CurrentCulture, out decValue);
                if (propertiesChanged != null)
                {
                    propertiesChanged(new propertiesChangedEventArgs
                    {
                        propName = "Total Income",
                        newValue = decValue
                    });
                }
            }
        }


        private Household _houseHold = new Household();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Household"), Editor(typeof(TypeEditor3), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Household houseHold
        {
            get { return _houseHold; }
            set { _houseHold = value; }
        }

        private Transport _transport = new Transport();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Transport"), Editor(typeof(TypeEditor8), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Transport transport
        {
            get { return _transport; }
            set { _transport = value; }
        }

        private Finance _finance = new Finance();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Finance"), Editor(typeof(TypeEditor2), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Finance finance
        {
            get { return _finance; }
            set { _finance = value; }
        }

        private Leisure _leisure = new Leisure();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Leisure"), Editor(typeof(TypeEditor5), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Leisure leisure
        {
            get { return _leisure; }
            set { _leisure = value; }
        }

        private RegularBills _regularBills = new RegularBills();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Regular Bills"), Editor(typeof(TypeEditor7), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public RegularBills regularBills
        {
            get { return _regularBills; }
            set { _regularBills = value; }
        }

        private Insurance _insurance = new Insurance();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Insurance"), Editor(typeof(TypeEditor4), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Insurance insurance
        {
            get { return _insurance; }
            set { _insurance = value; }
        }

        private Children _children = new Children();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Children"), Editor(typeof(TypeEditor1), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public Children children
        {
            get { return _children; }
            set
            {
                _children = value;
                System.Diagnostics.Debugger.Break();
            }
        }

        private OtherBills _otherBills = new OtherBills();
        [CategoryAttribute("(Monthly Outgoings)"), DisplayName("Other Bills"), Editor(typeof(TypeEditor6), typeof(UITypeEditor)), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public OtherBills otherBills
        {
            get { return _otherBills; }
            set { _otherBills = value; }
        }

        private void expandable_propertiesChanged(propertiesChangedEventArgs e)
        {
            if (propertiesChanged != null)
            {
                propertiesChanged(new propertiesChangedEventArgs
                {
                    propName = e.propName,
                    newValue = e.newValue
                });
            }
        }

    }
}

File: Household Budget Calculator/[VB]-Household Budget Calculator/VB/Household budget calculator/properties.vb
Imports System.ComponentModel
Imports System.Globalization
Imports System.Drawing.Design

<System.Serializable()> Public Class properties

    Public Event propertiesChanged(ByVal e As propertiesChangedEventArgs)

    Public Sub New()
        removeEventHandlers()
        addEventHandlers()
    End Sub

    Public Sub addEventHandlers()
        AddHandler _children.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _finance.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _houseHold.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _insurance.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _leisure.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _otherBills.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _regularBills.propertiesChanged, AddressOf expandable_propertiesChanged
        AddHandler _transport.propertiesChanged, AddressOf expandable_propertiesChanged
    End Sub

    Public Sub removeEventHandlers()
        RemoveHandler _children.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _finance.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _houseHold.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _insurance.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _leisure.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _otherBills.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _regularBills.propertiesChanged, AddressOf expandable_propertiesChanged
        RemoveHandler _transport.propertiesChanged, AddressOf expandable_propertiesChanged
    End Sub

    Private _net As Decimal
    <CategoryAttribute("(Monthly Income)"), _
    RefreshProperties(RefreshProperties.All), _
    DisplayName("Net monthly pay")> _
    Public Property net() As Decimal
        Get
            Return _net
        End Get
        Set(ByVal value As Decimal)
            _net = value
            totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2")
        End Set
    End Property

    Private _pension As Decimal
    <CategoryAttribute("(Monthly Income)"), _
    RefreshProperties(RefreshProperties.All), _
    DisplayName("Pension")> _
    Public Property pension() As Decimal
        Get
            Return _pension
        End Get
        Set(ByVal value As Decimal)
            _pension = value
            totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2")
        End Set
    End Property

    Private _otherIncome1 As Decimal
    <CategoryAttribute("(Monthly Income)"), _
    RefreshProperties(RefreshProperties.All), _
    DisplayName("Income from Savings + Investments")> _
    Public Property otherIncome1() As Decimal
        Get
            Return _otherIncome1
        End Get
        Set(ByVal value As Decimal)
            _otherIncome1 = value
            totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2")
        End Set
    End Property

    Private _otherIncome2 As Decimal
    <CategoryAttribute("(Monthly Income)"), _
    RefreshProperties(RefreshProperties.All), _
    DisplayName("Other Income")> _
    Public Property otherIncome2() As Decimal
        Get
            Return _otherIncome2
        End Get
        Set(ByVal value As Decimal)
            _otherIncome2 = value
            totalIncome = (net + pension + otherIncome1 + otherIncome2).ToString("c2")
        End Set
    End Property

    Private _totalIncome As String
    <CategoryAttribute("(Monthly Income)"), _
    RefreshProperties(RefreshProperties.All), _
    ReadOnlyAttribute(True), _
    DisplayName("Total Income"), _
    Editor(GetType(TypeEditor0), GetType(UITypeEditor))> _
    Public Property totalIncome() As String
        Get
            Return _totalIncome
        End Get
        Set(ByVal value As String)
            _totalIncome = value
            Dim decValue As Decimal
            Decimal.TryParse(value, NumberStyles.Currency, CultureInfo.CurrentCulture, decValue)
            RaiseEvent propertiesChanged(New propertiesChangedEventArgs With {.propName = "Total Income", .newValue = decValue})
        End Set
    End Property


    Private _houseHold As New Household
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Household"), _
    Editor(GetType(TypeEditor3), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property houseHold() As Household
        Get
            Return _houseHold
        End Get
        Set(ByVal value As Household)
            _houseHold = value
        End Set
    End Property

    Private _transport As New Transport
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Transport"), _
    Editor(GetType(TypeEditor8), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property transport() As Transport
        Get
            Return _transport
        End Get
        Set(ByVal value As Transport)
            _transport = value
        End Set
    End Property

    Private _finance As New Finance
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Finance"), _
    Editor(GetType(TypeEditor2), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property finance() As Finance
        Get
            Return _finance
        End Get
        Set(ByVal value As Finance)
            _finance = value
        End Set
    End Property

    Private _leisure As New Leisure
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Leisure"), _
    Editor(GetType(TypeEditor5), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property leisure() As Leisure
        Get
            Return _leisure
        End Get
        Set(ByVal value As Leisure)
            _leisure = value
        End Set
    End Property

    Private _regularBills As New RegularBills
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Regular Bills"), _
    Editor(GetType(TypeEditor7), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property regularBills() As RegularBills
        Get
            Return _regularBills
        End Get
        Set(ByVal value As RegularBills)
            _regularBills = value
        End Set
    End Property

    Private _insurance As New Insurance
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Insurance"), _
    Editor(GetType(TypeEditor4), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property insurance() As Insurance
        Get
            Return _insurance
        End Get
        Set(ByVal value As Insurance)
            _insurance = value
        End Set
    End Property

    Private _children As New Children
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Children"), _
    Editor(GetType(TypeEditor1), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property children() As Children
        Get
            Return _children
        End Get
        Set(ByVal value As Children)
            _children = value
            Stop
        End Set
    End Property

    Private _otherBills As New OtherBills
    <CategoryAttribute("(Monthly Outgoings)"), _
    DisplayName("Other Bills"), _
    Editor(GetType(TypeEditor6), GetType(UITypeEditor)), _
    DesignerSerializationVisibility(DesignerSerializationVisibility.Content)> _
    Public Property otherBills() As OtherBills
        Get
            Return _otherBills
        End Get
        Set(ByVal value As OtherBills)
            _otherBills = value
        End Set
    End Property

    Private Sub expandable_propertiesChanged(ByVal e As propertiesChangedEventArgs) 
        RaiseEvent propertiesChanged(New propertiesChangedEventArgs With {.propName = e.propName, .newValue = e.newValue})
    End Sub

End Class
